### ISSUE TITLE
Custom code module, injected globally

### ISSUE THREAD

**Issue by Jamiewarb** [hex-digital/lucidity-next-sanity-starter#11]
#Overview

Ability for CMS users to add code in the CMS to the header or footer of the site.

- [ ] Area in the CMS settings for Header code and Footer code
- [ ] Code comes through initial Site Manifest call, and is outputted in `<head>` or just before `</body>`
- [ ] Code areas should have syntax highlighting in CMS

**Comment by Jamiewarb** [comment]
The area for this doesn't yet exist in the CMS, but would likely go inside of a Site Settings singleton of sorts

**Comment by liamb13** [comment]
Hard a very quick play and found [react-textarea-code-editor](https://github.com/uiwjs/react-textarea-code-editor) to function quite nicely. 

Let me know your thoughts and I'll have a go at implementing it.

**Comment by Jamiewarb** [comment]
I like it, it looks powerful.

I've previously used this for basic code inputs: https://www.sanity.io/plugins/code-input

This code input has line numbering and syntax highlighting, but nothing else fancy like auto-indents.

I think the target audience for this is most likely marketers who are pasting in scripts from third-party tools like Hotjar, so may not need the extra functionality.

What do you think?

**Comment by Jamiewarb** [comment]
I'm thinking we create a Singleton for Custom Code, and pop it in the CMS Structure as Site Config > Custom Code. This then has a "Inside <head> tag" field, and a "Before </body> tag" field, and a short description to explain what kinds of things go there.

So we'd need to create a new Singleton document. I have all the Singleton stuff merged today, so I can set that up quick, and then we just need:

- A new singleton constant for Custom Code (`/packages/common/constants/schemaTypes.ts`)
- A new singleton schema type for Custom Code (`/apps/sanity/schema/types/singletons.ts`)
- Adding it to the site structure using the `singletonListItem` function (`/apps/sanity/structure/index.ts`)

Here's an example of what Webflow's Interface looks like for this

![image.png](https://volta.s3.fr-par.scw.cloud/image_01d88e1d5c.png)

**Comment by Jamiewarb** [comment]
Oh it looks like that `code-input` plugin is already in the Sanity app too—I'm using it for the "Raw HTML" modular block too

**Comment by liamb13** [comment]
> I think the target audience for this is most likely marketers who are pasting in scripts from third-party tools like Hotjar, so may not need the extra functionality.

Yeah this was who I thought it might be for. I went looking for the most textarea like format, thinking those head/footer code settings typically didn't look like a code editor. Seeing the WebFlow example has proven that wrong. 

That code-input one will be 10x easier to implement too! Why reinvent the wheel? 🤦🏻‍♂️

**Comment by Jamiewarb** [comment]
The Site Config area now exists within the sanity structure, so this singleton can go in there
